OPAL/br/src/main/scala/org/opalj/br/cfg/CFG.scala

```diff
@@ -722,6 +728,94 @@ case class CFG[I <: AnyRef, C <: CodeSequence[I]](
         )
     }
 
+    // We use this variable for caching, as the loop information of a CFG are permanent and do not
+    // need to be recomputed (see findNaturalLoops for usage)
+    private var naturalLoops: Option[List[List[Int]]] = None
+
+    /**
+     * ''findNaturalLoops'' finds all natural loops in this dominator tree and returns them as a
+     * list of lists where each inner list represents one loop and the Int values correspond to the
+     * indices of the nodes.
+     *
+     * @return Returns all found loops. The structure of the inner lists is as follows: The first
+     *         element of each inner list, i.e., each loop, is the loop header and the very last
+     *         element is the one that has a back-edge to the loop header. In between, elements are
+     *         ordered according to their occurrences, i.e., if ''n1'' is executed before ''n2'',
+     *         the index of ''n1'' is less than the index of ''n2''.
+     * @note This function only focuses on natural loops, i.e., it may / will produce incorrect
+     *       results on irreducible loops. For further information, see
+     *       [[http://www.cs.princeton.edu/courses/archive/spring03/cs320/notes/loops.pdf]].
+     */
+    def findNaturalLoops(): List[List[Int]] = {
```

Comment:
@errt or @TorunR, could you review that method?

Reply:
Outdated since handling complex control structures is not relevant to this PR anymore.